src/util/taproot.rs

```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    pub fn merkle_root(&self) -> Option<TapBranchHash> {
+        self.merkle_root
+    }
+
+    /// Output key(the key used in script pubkey) from Spend data. See also
+    /// [`TaprootSpendInfo::output_key_parity`]
+    pub fn output_key(&self) -> schnorr::PublicKey {
+        self.output_key
+    }
+
+    /// Parity of the output key. See also [`TaprootSpendInfo::output_key`]
+    pub fn output_key_parity(&self) -> bool {
+        self.output_key_parity
+    }
+
+    // Internal function to compute [`TaprootSpendInfo`] from NodeInfo
+    fn from_node_info<C: secp256k1::Verification>(
+        secp: &Secp256k1<C>,
+        internal_key: schnorr::PublicKey,
+        node: NodeInfo,
+    ) -> TaprootSpendInfo {
+        // Create as if it is a key spend path with the given merkle root
+        let root_hash = Some(TapBranchHash::from_inner(node.hash.into_inner()));
+        let mut info = TaprootSpendInfo::new_key_spend(secp, internal_key, root_hash);
+        for leaves in node.leaves {
+            let key = (leaves.script, leaves.ver);
+            let value = leaves.merkle_branch;
+            match info.script_map.get_mut(&key) {
+                Some(set) => {
+                    set.insert(value);
+                    continue; // NLL fix
+                }
+                None => {}
+            }
+            let mut set = BTreeSet::new();
+            set.insert(value);
+            info.script_map.insert(key, set);
+        }
+        info
+    }
+
+    /// Access the internal script map
+    pub fn as_script_map(&self) -> &ScriptMerkleProofMap {
+        &self.script_map
+    }
+
+    /// Obtain a [`ControlBlock`] for particular script with the given version.
+    /// Returns [`None`] if the script is not contained in the [`TaprootSpendInfo`]
+    /// If there are multiple ControlBlocks possible, this returns the shortest one.
+    pub fn control_block(&self, script_ver: &(Script, LeafVersion)) -> Option<ControlBlock> {
+        let merkle_branch_set = self.script_map.get(script_ver)?;
+        // Choose the smallest one amongst the multiple script maps
+        let smallest = merkle_branch_set
+            .iter()
+            .min_by(|x, y| x.0.len().cmp(&y.0.len()))
+            .expect("Invariant: Script map key must contain non-empty set value");
+        Some(ControlBlock {
+            internal_key: self.internal_key,
+            output_key_parity: self.output_key_parity,
+            leaf_version: LeafVersion::default(),
+            merkle_branch: smallest.clone(),
+        })
+    }
+}
+
+/// Builder for building taproot iteratively. Users can specify tap leaf or omitted/hidden
+/// branches in a DFS(Depth first search) walk to construct this tree.
+// Similar to Taproot Builder in bitcoin core
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TaprootBuilder {
+    // The following doc-comment is from bitcoin core, but modified for rust
+    // The current state of the builder.
```

Comment:
Can't parse the meaning of this line of the comment and how it is related to the previous line (the same sentence or a new one)
```suggestion
```